src/storage/btree.c

```diff
@@ -14404,9 +14427,25 @@ btree_find_boundary_leaf (THREAD_ENTRY * thread_p, BTID * btid, VPID * pg_vpid,
 
       btree_read_fixed_portion_of_non_leaf_record (&rec, &nleaf);
       C_vpid = nleaf.pnt;
-      C_page = pgbuf_fix (thread_p, &C_vpid, OLD_PAGE, PGBUF_LATCH_READ, PGBUF_UNCONDITIONAL_LATCH);
+      C_page = pgbuf_fix_read_old_and_check_repl_desync (thread_p, C_vpid, PGBUF_UNCONDITIONAL_LATCH);
       if (C_page == NULL)
 	{
+#ifdef SERVER_MODE
+	  if (er_errid () == ER_PAGE_AHEAD_OF_REPLICATION)
+	    {
+	      pgbuf_unfix (thread_p, P_page);
+
+	      //wait for replication
+	      passive_tran_server *const pts_ptr = get_passive_tran_server_ptr ();
+	      const LOG_TDES *const tdes = LOG_FIND_CURRENT_TDES (thread_p);
+	      assert (tdes->page_desync_lsa.is_null ());
+	      pts_ptr->wait_replication_past_target_lsa (tdes->page_desync_lsa);
+
+	      //reset the search
+	      first_time = true;
+	      continue;
+	    }
+#endif
```

Comment:
The `pgbuf_fix` in the `again:` goto section shouldn't be replaced with `pgbuf_fix_read_old_and_check_repl_desync`?
I understand that the `again` section is executed if the boundady leaf page of the btree does not contain any keys - `key_cnt` is 0.
In light of this, I suggest that:
- the initial version of this function should simply be modified to bail out when a page desync is found (return null page with a an extra status - signature might need to change) - maybe change name to `btree_find_boundary_leaf_with_repl_desync_check`
- a wrapper function with the old name `btree_find_boundary_leaf` be put in place that does a loop call of this function with check whether replication needs to be waited